### DISCUSSION TITLE
Compile-time assert

### DISCUSSION THREAD

**Discussion by nordlow** [vlang/v#11851] [Ideas]
D has `static assert`. C++ has `static_assert`.

V should have `$assert`.

**Comment by dumblob** [comment]
Could you maybe show some use cases? *V*'s type system and libraries are built in a way which avoids the need for a static assert (that's why I'm asking for some good use cases :wink:).

**Comment by nordlow** [comment]
Search for `static assert` under dlang as

1. https://github.com/dlang/phobos/search?q=%22static+assert%22
2. https://github.com/dlang/druntime/search?q=%22static+assert%22
3. https://github.com/dlang/dmd/search?q=%22static+assert%22

In short; (pure function) tests that can be executed at compile-time are run at compile-time using `static assert` for faster feedback in the CI loop.

**Comment by nordlow** [comment]
> Could you maybe show some use cases? _V_'s type system and libraries are built in a way which avoids the need for a static assert (that's why I'm asking for some good use cases wink).

Would be interesting to see some details on how you plan to avoid the need for compile-time asserts. Especially in a generic context.

**Comment by dumblob** [comment]
> Search for `static assert` under dlang as
> 
> 1. https://github.com/dlang/phobos/search?q=%22static+assert%22
> 2. https://github.com/dlang/druntime/search?q=%22static+assert%22
> 3. https://github.com/dlang/dmd/search?q=%22static+assert%22
> 
> In short; (pure function) tests that can be executed at compile-time are run at compile-time using `static assert` for faster feedback in the CI loop.

Does this hold also for *V*? *V* is so fast, that I didn't see any need for compile-time tests yet (it's always possible to write a runtime test for compile-time functionality). And assuming (hoping :wink:) it'll stay fast, there won't be such need IMHO. Or do you have anything specific in mind? Any **concrete** examples of (potential) **V code**?

> Would be interesting to see some details on how you plan to avoid the need for compile-time asserts. Especially in a generic context.

I'm confused - what need do you mean? *V*'s generics generate specialized C code for every **used** combination. Where would I need any compile-time code (and thus static assert) in this regard (assuming *V*'s compilation stays fast and thus all potentially "compile time" tests can be realized in run time)?

**Comment by nordlow** [comment]
It's not about performance, it's about proving correctness at compile-time removing the need for _any_ run-time tests whatsoever.

Ok, I'll give you a very concrete example. Take something a simple as filtered array concatenation. Does V allow an array of ints `y` to be appended to an array of longs `x`? If not, would you regard that as a simple performant language? Now take this to the next level. 

`x.append(y.filter(e => e != n))`

without requiring any intermediate array storage of the result of `y.filter(e => e != n)`. And of course for any conceivable combination of element types of `x` and `y` that produces a relevant/correct result.

Can this be done in V guaranteeing no run-time errors?

If not, would you still regard that as a simple high-performance language?

V is very keen on selling itself as a language that avoid unnecessary (intermediate) allocations. That statement doesn't hold for these situations.

**Comment by dumblob** [comment]
> It's not about performance, it's about proving correctness at compile-time removing the need for _any_ run-time tests whatsoever.

As said above - *V*'s generics are just a specialized struct (or function) in C. Correctness is thus guaranteed on two levels - first is *V* and in case it doesn't catch it (which should not happen), then there is still the C type check. So assuming *V 1.0* (i.e. a stable version of *V*) there is no way to compile code using mutually incompatible types or operations or whatever (see promotions and "nominal" type comparison below).

I think thought what you're trying to convey is what's being usally called "structural types vs nominal types" dichotomy. *V*'s generics are AFAIK purely structural and thus if the structure of the given *generic type* (e.g. `T`) matches the algorithm (not the generic interface which is just an umbrella for syntactic description, but is void in a specialized/lowered AST), *V* simply generates the C code and tries to compile it. I'd call this "structurally-focused". It's a bit less restrictive than nominally-focused generics which seems to be what D does. In a nominal type system you just get two canonicalized types (imagine `/usertypes/mymod1/submod1/MyStruct/MyAnotherStruct/int` and `/builtintypes/base/int`) and you need to prove **in compile time** (ideally before you lower anything into the specialized/lowered AST) that both types are either equivalent or "compatible" (I deliberately not say "castable" or "coerce" or ...) which by it's virtue needs quite often manual help (unless you have a [language with inductive types and a formal proving engine]( https://github.com/uwu-tech/Kind )) - including static assert etc.

In practice though type systems are a combination (at least in widespread languages and not purely functional languages which tend to stick with nominal types only, and not duck-typing only languages which tend to stick with structural types only). *V* currently offers both, but emphasis is on structural typing IMHO. You seem to be wanting this to be shifted more into the nominal typing spectrum.

A note aside - from the above it'd seem structural typing is generally strictly better than nominal typing. That's true when it comes to its proliferation in all the written code (disregarding the language) on the planet. But as of few months ago (when I checked last), the latest research did not have any clue how to make a general (in terms of "generic", but that word would be convoluted in this discussion :wink:) structural typing system **decidable** (there are only constrained structural typing systems known to be proved decidable). Unlike general nominal typing systems which were proved decidable.

So far nobody AFAIK tried to prove that current *V*'s type system is decidable. But I personally have the feeling it is (at least its architecture/semantics - i.e. after **all** - even currently unknown - bugs get fixed) - i.e. it's probably not a general structural typing system.

> Ok, I'll give you a very concrete example.

Perfect, thanks!

>Take something simple as filtered array concatenation. Does V allow an array of ints `y` to be appended to an array of longs `x`?

Sure, why not? It seems to be in accordance with existing [type promotions and the need to explicitly manually cast everything else](https://github.com/vlang/v/blob/master/doc/docs.md#types ) - i.e. everything which is not *nominally* (yes, *V* is also a mixture as I said above :wink:) the same type.

>If not, would you regard that as a simple performant language? Now take this to the next level.
> 
> `x.append(y.filter(e => e != n))`
> 
> without requiring any intermediate array storage of the result of `y.filter(e => e != n)`. And of course for any conceivable combination of element types of `x` and `y` that produces a relevant/correct result.
> 
> Can this be done in V guaranteeing no run-time errors?

Yes. Why would you expect any run-time errors? As I said above - the lowered AST can be checked (*V* has the information to check it). If it's not currently being checked, it's a bug **in the implementation** (not in the language itself) and shall be fixed before *V 1.0*.

Though I'm not sure about the (irrelevant - as it's an implementation detail) intermediate array construction behind the scenes as you'd probably need to use slicing and not a traditional array - but [this is all in flux](https://github.com/vlang/v/issues/5605 ) and I'm not up to date with all these details.

> If not, would you still regard that as a simple high-performance language?

No, that'd be a pity, of course.

**Comment by nordlow** [comment]
> Take something simple as filtered array concatenation. Does V allow an array of ints y to be appended to an array of longs x?

I was thinking about a builtin concatenation function. This can of course be implemented as a generic function.

Instead lets focus on a variant

```
x.append(y.filterLazy(e => e != n))
```

. For that to be implemented in V you need to have `filterLazy` returning a D-style lazy range (Rust-like iterator).

Are such lazy algorithms planned? A problem here is that the name `map` and `filter` are in V already reserved for eager evaluation meaning that for backwards consistency, V would have to implement two namings for every upcoming algorithm, the eager `do_something` and the lazy `do_something_lazy`.

D solves this in an orthgonal way by having only lazy-algorithm version and when an array result is needed you do

    auto resultArray = x.doSomething.array;

where `array` is defined in `std.array` and doSomething is typically a lazy algorithm. if x.doSomething is already an `array` then `.array` is a no-op.

This does not only cut code-duplication in two (compared to how current state of things in V) but also makes it conceptually easier to write and read code in D.

**Comment by dumblob** [comment]
>Are such lazy algorithms planned?

No ideas - others should answer :wink:.

But I for myself hope the "views" (e.g. slices without copying or whatever) will be enough to not change any current interfaces (i.e. there will not be the need to introduce additional methods like `filterLazy`). I think this better aligns with "data abstraction over operational abstraction" principle.

**Comment by nordlow** [comment]
Array views/slices are not enough to express the (intermediate) results of all combinations of lazy calculations that is needed. Provided, you want V to get anywhere near the performance of C++, Rust and D for more advanced computations and not have to resort to C-style non-scalable ways of expressing algorithms.

**Comment by dumblob** [comment]
I'm not exactly sure what you're referring to, but I think Go proves, that the "non-scalable ways of expressing algorithms" are highly scalable in practice. Could you maybe elaborate?

Or instead of lazy calculations you meant rather something more like "tight interaction among generics"? Because none of the 3 languages you named are capable of expressing what I'd expect from "interaction among generics" - namely something like https://github.com/zero-functional/zero-functional .

Any concrete examples you meant (after correlating them with Go libraries for the same kind of computation)?

**Comment by nordlow** [comment]
> I'm not exactly sure what you're referring to, but I think Go proves, that the "non-scalable ways of expressing algorithms" are highly scalable in practice.

So why is Go planning to add generics to the language [1]?

> Or instead of lazy calculations you meant rather something more like "tight interaction among generics"? Because none of the 3 languages you named are capable of expressing what I'd expect from "interaction among generics" - namely something like https://github.com/zero-functional/zero-functional.

What is meant by zero-cost here? Cost of compiling or cost of running or both?

[1] https://blog.golang.org/generics-next-step

**Comment by dumblob** [comment]
> > I'm not exactly sure what you're referring to, but I think Go proves, that the "non-scalable ways of expressing algorithms" are highly scalable in practice.
> 
> So why is Go planning to add generics to the language [1]?

First, I'm talking about existing code bases. Second, those Go's generics as I've evaluated them 3 years ago seem to be still slightly less flexible than *V*'s. In that sense, I'd be again interested in a concrete example of what can't be expressed in *V* sufficiently well, so that it hinders any of: *runtime performance, compile time performance, code readability, correctness* by more than ~3% from a hypothetical *V* having compile-time assert.

> > Or instead of lazy calculations you meant rather something more like "tight interaction among generics"? Because none of the 3 languages you named are capable of expressing what I'd expect from "interaction among generics" - namely something like https://github.com/zero-functional/zero-functional.
> 
> What is meant by zero-cost here? Cost of compiling or cost of running or both?

I suppose they mean the runtime cost of abstractions used (ephemeral object/struct instances & functions/lambdas both have some cost). But cost of compilation will not be much worse from how I know Nim's metaprogramming.

**Comment by Cons-Cat** [comment]
I use `static_assert()` all the time in my C++ functions to help express that `constexpr` + templated functions are being used correctly, because otherwise the GNU error messages get ridiculously esoteric compared to what I can write when predicting specific misuse-cases.

Between V's error messages being generally better than C++ compilers, and the language just being fundamentally harder to misuse due to its general simplicity, I'm not sure if it would benefit from an `$assert` to the same degree as other langs with more complex / powerful metaprogramming. Especially not considering some of the other solutions like generic constraints that have been discussed to prevent misusing generics.

I think the gg module, which has a lot of runtime asserts currently, might benefit from something like this, but maybe not to a large degree since it would still need the runtime asserts _anyways_. Maybe asserts could evalute at compile time _and_ runtime where possible?

```v
fn main() {
    assert(false)
    return
}
```
This compiles, but panics when running. We can trivially prove by reading the code that it isn't useful. Can a compiler do the same?

This is just one angle, though. Other use cases for a static assert exist, ofc.
